### ISSUE TITLE
Prance is not reading ref for encoded url

### ISSUE THREAD

**Issue by krishnajangid** [jfinkhaeuser/prance#56]
## Expected Behaviour
Swagger reads ref file with encoded url but prance through errr file not found error.

### Minimal Example Spec
$ref:'#/paths/~1api~1%7Bname%7D~1data

## Actual Behaviour
Error: 
Cannot resolve reference "file:///<dir name>/paths/~1api~1%7Bname%7D~1data
## Environment
- OS: 
- Python version: 
- Swagger/OpenAPI version used:
- Backend: (flex, swagger-spec-validator or openapi-spec-validator)

@jfinkhaeuser

**Comment by jfinkhaeuser** [comment]
I'll have to look into that. Thanks for reporting!

**Comment by krishnajangid** [comment]
Can you please provide me an example to iterate through  YAML spec
validation error.

**Comment by krishnajangid** [comment]
@jfinkhaeuser please look into it. merge it https://github.com/jfinkhaeuser/prance/pull/57/files

**Comment by jfinkhaeuser** [comment]
I've reviewed #57, and came to the conclusion it's not the right way to solve the issue.

In the `reference_iterator` implementation, we don't know whether we've got a reference to some other file, or perhaps some HTTP URL. In the case of an HTTP URL, removing URL encoding here would be the wrong choice. It's likely that we'd have to URL-encode everything again before sending it to a server.

If you look at https://github.com/jfinkhaeuser/prance/blob/master/prance/util/url.py#L128, when we're about to use the reference, we're normalizing it by replacing `~1` and `~0` as documented in the JSON pointer specs. One could argue that this should be in the iterator where you tried to the decode the URL, because it's specific to the JSON pointer syntax.

This function is called from the `RefResolver`, which uses the `url` submodule again to fetch the resource. In particular, https://github.com/jfinkhaeuser/prance/blob/master/prance/util/url.py#L137 does the actual fetching and looks at the URL scheme.

For local files, it uses the `fs` submodule, specifically https://github.com/jfinkhaeuser/prance/blob/master/prance/util/fs.py#L275 - if the change should be made anywhere, then there. I'm not convinced, though.

The effect that this would have is that you *have* to have a file system capable of naming files with some encoding of unicode. There are still plenty of situations where things aren't so simple. That's why on most operating systems, file systems simply pass the bytes of filenames through, and don't try to interpret them as any encoding. The exception here would be Windows, where the Win32 APIs place some requirements on you.

The TL;DR is, it could do more harm than good.

I would propose a change to the `fs` module instead, where instead of calling `open()/io.open()`, we call a wrapper function that tries to read first the urldecoded file name (local files are more likely like that, in my opinion), and on failure tries the original, urlencoded version.

WDYT?

**Comment by krishnajangid** [comment]
I'll look into it. And let you know.

**Comment by krishnajangid** [comment]
I had tried to debug. The error that I found in another place. In the path module. The **path_get** gets when it trying to get an example from a different place.

For that path key error was occurring because it was encoded.
So to resolve that issue. We will try to get value form dict if a key error occurs then try to decode it and get value for that key. I had made changes for the same pull request(#57). Please go through it.

For the YAML data, I was trying to parse. I had mailed you. So plz, try to run for that.

And as you mention about open file in fs module. I had tried for one example. It was working fine if we do chnages what I did (#57)
We might need to check for different senario.

**Comment by jfinkhaeuser** [comment]
Thank you for sending the example. This and re-reading RFCs (and YAML specs) have led me to believe that this isn't an issue in prance.

[YAML](https://yaml.org/spec/1.2/spec.html) positions itself as a superset of [JSON](https://www.ietf.org/rfc/rfc4627.txt), and in particular they share much the same string encoding rules. That is, quoted strings can contain just about any unicode character, escaped or unescaped - only a few have to be escaped.

The escaping rules in JSON follow C conventions: "\u002f" is a forward slash. In YAML, it's a bit more complex based on whether the code point can be represented in 8-bit, 16-bit or 32-bit, but for our purposes, it's the same: URL-escaping with a `%7B` hex code is *not* interpreted as unicode characters, but as the individual characters `%`, `7` and `B`.

In other words, URL-decoding JSON references would be a bug.

Which means your YAML is simply wrong. The fix is simply to replace the %-encoded characters with their unencoded counterpart:

```diff
--- demo.yaml	2020-01-20 10:31:37.645485060 +0100
+++ fixed.yaml	2020-01-20 10:48:41.792971464 +0100
@@ -34,7 +34,7 @@
                     error: false
                     message: OK
                 Error:
-                  $ref: '#/paths/~1pet~1%7Bid%7D~1/get/responses/200/content/application~1json/examples/Pet%20-%20Error'
+                  $ref: '#/paths/~1pet~1{id}~1/get/responses/200/content/application~1json/examples/Pet - Error'
   '/pet/{id}/':
     get:
       tags:
```

**Comment by krishnajangid** [comment]
Thanks, @jfinkhaeuser  for the response. That file generated using the CLI library. I used that in swagger editor it was working fine. And tested with a few other libraries like spectral(https://stoplight.io/open-source/spectral)  it worked fine in that. As you mention it is not the right way to write YAML. I strongly believe in that. So if you feel that's the wrong issue then we can ignore it.

**Comment by jfinkhaeuser** [comment]
I have no doubt that other implementations do this wrong. And maybe there's something somewhere that I've missed in the specs. But so far, I can't find it.

**Comment by krishnajangid** [comment]
Okay, make sense. Thanks
